File: src/game/production/recipes.ts
import { configNumber } from '../activity/gathering'
import type { RecipeRow } from '../data/recipeTypes'
import type { ActivityRow, GameDatabase } from '../data/types'
import { canKnowRecipe as recipeKnown } from '../recipes/knowledge'
import type { PlayerSave } from '../save/types'
import { requirementsForEntity } from '../activity/requirements'

export interface RecipeIngredient {
  itemId: string
  quantity: number
}

export function isCompleteRecipe(recipe: RecipeRow): boolean {
  if (recipe.Status === 'Needs Data') return false
  if (typeof recipe['Base Duration Seconds'] !== 'number') return false
  if (typeof recipe['XP Reward'] !== 'number') return false
  if (typeof recipe['Output Quantity'] !== 'number') return false
  if (!recipe['Output Item ID'] || !recipe['Facility ID'] || !recipe['Skill ID']) return false
  if (typeof recipe['Proficiency Level'] !== 'number') return false
  return true
}

export function recipeIngredients(recipe: RecipeRow): RecipeIngredient[] {
  const out: RecipeIngredient[] = []
  const pairs: Array<[string | null, number | null]> = [
    [recipe['Ingredient 1 Item ID'], recipe['Ingredient 1 Quantity']],
    [recipe['Ingredient 2 Item ID'], recipe['Ingredient 2 Quantity']],
    [recipe['Ingredient 3 Item ID'], recipe['Ingredient 3 Quantity']],
  ]
  for (const [itemId, quantity] of pairs) {
    if (itemId && typeof quantity === 'number' && quantity > 0) {
      out.push({ itemId, quantity })
    }
  }
  return out
}

export function inventoryCount(save: PlayerSave, itemId: string): number {
  return save.inventory.find((stack) => stack.itemId === itemId)?.quantity ?? 0
}

export function maxCraftsFromMaterials(save: PlayerSave, recipe: RecipeRow): number {
  const ingredients = recipeIngredients(recipe)
  if (ingredients.length === 0) return 0
  let max = Number.POSITIVE_INFINITY
  for (const ingredient of ingredients) {
    max = Math.min(max, Math.floor(inventoryCount(save, ingredient.itemId) / ingredient.quantity))
  }
  return Number.isFinite(max) ? Math.max(0, max) : 0
}

export function queueCapSeconds(db: GameDatabase): number {
  return configNumber(db, 'standard_production_queue_cap', 24) * 3600
}

export function maxCraftsFromQueueCap(db: GameDatabase, recipe: RecipeRow): number {
  const duration = Math.max(1, recipe['Base Duration Seconds'])
  return Math.max(0, Math.floor(queueCapSeconds(db) / duration))
}

/** Hard proficiency gate + knowledge-source unlocks. */
export function canKnowRecipe(save: PlayerSave, db: GameDatabase, recipe: RecipeRow): boolean {
  return recipeKnown(save, db, recipe)
}

/** Castle kitchen shares the Town kitchen recipe book. */
const SHARED_RECIPE_FACILITY_IDS: Record<string, string> = {
  'FAC-0010': 'FAC-0001',
}

export function recipeFacilityIdForLookup(facilityId: string): string {
  return SHARED_RECIPE_FACILITY_IDS[facilityId] ?? facilityId
}

export function recipeMatchesFacility(recipeFacilityId: string, activityFacilityId: string): boolean {
  return recipeFacilityId === recipeFacilityIdForLookup(activityFacilityId)
}

export function facilityIdForActivity(db: GameDatabase, activityId: string): string | null {
  const station = requirementsForEntity(db, 'Activity', activityId).find(
    (row) => row['Requirement Type'] === 'Station',
  )
  const value = station?.['Reference ID / Value']
  return typeof value === 'string' && value.length > 0 ? value : null
}

export function isStandardProductionActivity(db: GameDatabase, activity: ActivityRow): boolean {
  if (activity['Pool ID']) return false
  return facilityIdForActivity(db, activity['Activity ID']) != null
}

export function recipesForActivity(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
): RecipeRow[] {
  const facilityId = facilityIdForActivity(db, activityId)
  if (!facilityId) return []
  return db.Recipes.filter(
    (recipe) =>
      isCompleteRecipe(recipe) &&
      recipeMatchesFacility(recipe['Facility ID'], facilityId) &&
      canKnowRecipe(save, db, recipe),
  ).sort((a, b) => a['Proficiency Level'] - b['Proficiency Level'])
}

export function getRecipe(db: GameDatabase, recipeId: string): RecipeRow | undefined {
  return db.Recipes.find((recipe) => recipe['Recipe ID'] === recipeId)
}

export function clampProductionQuantity(
  db: GameDatabase,
  save: PlayerSave,
  recipe: RecipeRow,
  requested: number,
): number {
  const wanted = Math.floor(requested)
  if (wanted <= 0) return 0
  return Math.min(
    wanted,
    maxCraftsFromMaterials(save, recipe),
    maxCraftsFromQueueCap(db, recipe),
  )
}

File: packages/ik_rules/lib/src/production/recipes.dart
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../activity/requirements.dart';
import '../config.dart';
import '../js_compat.dart';
import '../recipes/knowledge.dart';
import '../save/generated/save_models.dart';

/// One material line of a recipe.
class RecipeIngredient {
  const RecipeIngredient({required this.itemId, required this.quantity});

  final String itemId;
  final num quantity;

  Map<String, Object?> toJson() => <String, Object?>{'itemId': itemId, 'quantity': quantity};
}

bool isCompleteRecipe(RecipeRow recipe) {
  if (recipe.raw['Status'] == 'Needs Data') return false;
  if (recipe.raw['Base Duration Seconds'] is! num) return false;
  if (recipe.raw['XP Reward'] is! num) return false;
  if (recipe.raw['Output Quantity'] is! num) return false;
  if (isBlank(recipe.raw['Output Item ID'] as String?) ||
      isBlank(recipe.raw['Facility ID'] as String?) ||
      isBlank(recipe.raw['Skill ID'] as String?)) {
    return false;
  }
  return recipe.raw['Proficiency Level'] is num;
}

List<RecipeIngredient> recipeIngredients(RecipeRow recipe) {
  final out = <RecipeIngredient>[];
  for (var slot = 1; slot <= 3; slot += 1) {
    final itemId = recipe.raw['Ingredient $slot Item ID'];
    final quantity = recipe.raw['Ingredient $slot Quantity'];
    if (itemId is String && itemId.isNotEmpty && quantity is num && quantity > 0) {
      out.add(RecipeIngredient(itemId: itemId, quantity: quantity));
    }
  }
  return out;
}

num inventoryCount(PlayerSave save, String itemId) {
  return save.inventory.firstWhereOrNull((stack) => stack.itemId == itemId)?.quantity ?? 0;
}

num maxCraftsFromMaterials(PlayerSave save, RecipeRow recipe) {
  final ingredients = recipeIngredients(recipe);
  if (ingredients.isEmpty) return 0;
  num max = double.infinity;
  for (final ingredient in ingredients) {
    max = math.min(max, (inventoryCount(save, ingredient.itemId) / ingredient.quantity).floor());
  }
  return max.isFinite ? math.max(0, max) : 0;
}

num queueCapSeconds(GameDatabase db) {
  return configNumber(db, 'standard_production_queue_cap', 24) * 3600;
}

num maxCraftsFromQueueCap(GameDatabase db, RecipeRow recipe) {
  final duration = math.max(1, jsNumber(recipe.raw['Base Duration Seconds']));
  return math.max(0, (queueCapSeconds(db) / duration).floor());
}

/// Shared recipe books: castle/citadel stations reuse Town facility catalogs.
const Map<String, String> _sharedRecipeFacilityIds = <String, String>{
  'FAC-0010': 'FAC-0001',
  'FAC-0012': 'FAC-0001',
  'FAC-0013': 'FAC-0003',
  'FAC-0014': 'FAC-0004',
  'FAC-0015': 'FAC-0006',
};

/// Citadel special-production facilities reuse Town project catalogs.
const Map<String, String> sharedProjectFacilityIds = <String, String>{
  'FAC-0013': 'FAC-0003',
  'FAC-0016': 'FAC-0005',
};

String projectFacilityIdForLookup(String facilityId) {
  return sharedProjectFacilityIds[facilityId] ?? facilityId;
}

String recipeFacilityIdForLookup(String facilityId) {
  return _sharedRecipeFacilityIds[facilityId] ?? facilityId;
}

bool recipeMatchesFacility(String recipeFacilityId, String activityFacilityId) {
  return recipeFacilityId == recipeFacilityIdForLookup(activityFacilityId);
}

String? facilityIdForActivity(GameDatabase db, String activityId) {
  final station = requirementsForEntity(
    db,
    'Activity',
    activityId,
  ).firstWhereOrNull((row) => row.raw['Requirement Type'] == 'Station');
  final value = station?.raw['Reference ID / Value'];
  return value is String && value.isNotEmpty ? value : null;
}

bool isStandardProductionActivity(GameDatabase db, ActivityRow activity) {
  if (isNotBlank(activity.raw['Pool ID'] as String?)) return false;
  return facilityIdForActivity(db, jsString(activity.raw['Activity ID'])) != null;
}

List<RecipeRow> recipesForActivity(GameDatabase db, PlayerSave save, String activityId) {
  final facilityId = facilityIdForActivity(db, activityId);
  if (facilityId == null) return const <RecipeRow>[];
  final matching = db.recipes
      .where(
        (recipe) =>
            isCompleteRecipe(recipe) &&
            recipeMatchesFacility(jsString(recipe.raw['Facility ID']), facilityId) &&
            canKnowRecipe(save, db, recipe),
      )
      .toList();
  mergeSort(
    matching,
    compare: (a, b) {
      final delta = jsNumber(a.raw['Proficiency Level']) - jsNumber(b.raw['Proficiency Level']);
      return delta == 0 || delta.isNaN ? 0 : (delta > 0 ? 1 : -1);
    },
  );
  return matching;
}

RecipeRow? getRecipe(GameDatabase db, String recipeId) {
  return db.recipes.firstWhereOrNull((recipe) => recipe.raw['Recipe ID'] == recipeId);
}

num clampProductionQuantity(GameDatabase db, PlayerSave save, RecipeRow recipe, num requested) {
  final wanted = requested.floor();
  if (wanted <= 0) return 0;
  return math.min(
    wanted,
    math.min(maxCraftsFromMaterials(save, recipe), maxCraftsFromQueueCap(db, recipe)),
  );
}
